cmd/contact-auditor/main.go

```diff
@@ -0,0 +1,203 @@
+			FROM registrations AS r
+				INNER JOIN certificates AS c on c.registrationID = r.id
+			WHERE r.contact NOT IN ('[]', 'null')
+				AND c.expires >=`,
+			c.clk.Now().Add(-c.grace).Format("2006-01-02T15:04:05Z")),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error performing db query: %s", err)
+	}
+	return rows, nil
+}
+
+// run retrieves a cursor from `beginAuditQuery` and then audits the
+// `contact` column of all returned rows for abnormalities or policy
+// violations.
+func (c contactAuditor) run(resChan chan *result) error {
+	c.logger.Infof("Beginning database query")
+	rows, err := c.beginAuditQuery()
+	if err != nil {
+		return err
+	}
+
+	for rows.Next() {
+		var res result
+		err := rows.Scan(&res.ID, &res.Contact)
+		if err != nil {
+			return err
+		}
+
+		err = res.unmarshalContact()
+		if err != nil {
+			c.logger.Errf("Unmarshal failed for ID: %d due to: %s", res.ID, err)
+		}
+
+		err = res.validateContact()
+		if err != nil {
+			c.logger.Errf("Validation failed for ID: %s due to: %s", res.ID, err)
+		}
+
+		// Only used for testing.
+		if resChan != nil {
+			resChan <- &res
+		}
+	}
+	// Ensure the query wasn't interrupted before it could complete.
+	err = rows.Close()
+	if err != nil {
+		// Log an error but continue processing results.
+		c.logger.Errf("Query interrupted due to: %s", err)
+		queryInterrupted = true
+	} else {
+		c.logger.Info("Query completed successfully")
+	}
+
+	// Only used for testing.
+	if resChan != nil {
+		close(resChan)
+	}
+
+	return nil
+}
+
+func main() {
+	configFile := flag.String("config", "", "File containing a JSON config.")
+	grace := flag.Duration(
+		"grace", 2*24*time.Hour, "Include contacts of subscribers with certificates that expired <grace>\n period from now")
+	flag.Parse()
+
+	logger := cmd.NewLogger(cmd.SyslogConfig{StdoutLevel: 7})
+
+	configData, err := ioutil.ReadFile(*configFile)
+	cmd.FailOnError(err, fmt.Sprintf("Error reading file: %q", *configFile))
+
+	// Load JSON configuration.
+	type config struct {
+		ContactAuditor struct {
+			DB cmd.DBConfig
+			cmd.PasswordConfig
+		}
+	}
+
+	var cfg config
+	err = json.Unmarshal(configData, &cfg)
+	cmd.FailOnError(err, "Couldn't unmarshal config")
+
+	// Setup database client.
+	dbURL, err := cfg.ContactAuditor.DB.URL()
+	cmd.FailOnError(err, "Couldn't load DB URL")
+
+	db, err := sql.Open("mysql", fmt.Sprintf("%s?readTimeout=14s&writeTimeout=14s&timeout=1s", dbURL))
+	cmd.FailOnError(err, "Couldn't setup database client")
+
+	// Apply database settings.
+	db.SetMaxOpenConns(cfg.ContactAuditor.DB.MaxOpenConns)
+	db.SetMaxIdleConns(cfg.ContactAuditor.DB.MaxIdleConns)
+	db.SetConnMaxLifetime(cfg.ContactAuditor.DB.ConnMaxLifetime.Duration)
+	db.SetConnMaxIdleTime(cfg.ContactAuditor.DB.ConnMaxIdleTime.Duration)
+
+	// Setting isolation level to `READ UNCOMMITTED` improves
+	// performance at the cost of consistency.
+	_, err = db.Exec("SET SESSION TRANSACTION ISOLATION LEVEL READ UNCOMMITTED;")
```

Comment:
This is a really fun and surprising pitfall. Session variables like this are set on a per-connection basis. But Go's SQL driver allocations connections dynamically and on demand. So the fact that you ran this on one connection is no guarantee that your subsequent query is run on the same connection.

The fix is that the driver itself needs to know about certain things that you want to be set on every connection, before any transaction is run on it. You do that with parameters to the DSN: https://github.com/go-sql-driver/mysql/#system-variables. There's an example at that link of how to do this for isolation level.

That's probably the best way to do what you want here. For completeness' sake, another way you could achieve it is starting a transaction. When you do that, you get back a [`*sql.Tx`](https://golang.org/pkg/database/sql/#Tx) and execute queries on that. Those queries are guaranteed to all run on the same connection, because a transaction has to happen on a single connection.